network/ssh/testing/sshserver.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package testing
+
+import (
+	"net"
+	"time"
+
+	"github.com/juju/errors"
+	jc "github.com/juju/testing/checkers"
+	"golang.org/x/crypto/ssh"
+	gc "gopkg.in/check.v1"
+)
+
+// SSHKey1 generated with `ssh-keygen -b 256 -C test-only -t ecdsa -f test-key`
+var SSHKey1 = `-----BEGIN EC PRIVATE KEY-----
+MHcCAQEEILhuaRN6CI4h85SjOFV2+SU1uslRirsyyhGdsVmkKaC2oAoGCCqGSM49
+AwEHoUQDQgAESKoQ2r2l3hdXf9K+j+KsTwpTHNWMdd7gsl0tgy+77DYbz7DUDml1
+vIBDwimK29kn9WpPU8WSW23ZFPLk53mNTw==
+-----END EC PRIVATE KEY-----
+`
+
+var SSHPub1 = "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEiqENq9pd4XV3/Svo/irE8KUxzVjHXe4LJdLYMvu+w2G8+w1A5pdbyAQ8IpitvZJ/VqT1PFkltt2RTy5Od5jU8= test-only"
+
+// SSHKey2 generated with `ssh-keygen -b 256 -C test-only -t ed25519 -f test-key`
+var SSHKey2 = `-----BEGIN OPENSSH PRIVATE KEY-----
+b3BlbnNzaC1rZXktdjEAAAAABG5vbmUAAAAEbm9uZQAAAAAAAAABAAAAMwAAAAtzc2gtZW
+QyNTUxOQAAACANk3iR1VrTsEfIyQDXkrZajtOIwmKBdz+hAN90VXdxOQAAAJDQ4EH60OBB
++gAAAAtzc2gtZWQyNTUxOQAAACANk3iR1VrTsEfIyQDXkrZajtOIwmKBdz+hAN90VXdxOQ
+AAAEB0Vb6XYd1aFm1dl+37KgqgEeZDuFRlSHjeHrXEDFP4Iw2TeJHVWtOwR8jJANeStlqO
+04jCYoF3P6EA33RVd3E5AAAACXRlc3Qtb25seQECAwQ=
+-----END OPENSSH PRIVATE KEY-----
+`
+
+var SSHPub2 = "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIA2TeJHVWtOwR8jJANeStlqO04jCYoF3P6EA33RVd3E5 test-only"
+
+// denyPublicKey implements the SSH PublicKeyCallback API, but just always
+// denies any public key it gets.
+func denyPublicKey(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+	return nil, errors.Errorf("public key denied")
+}
+
+// CreateSSHServer launches an SSH server that will use the described private
+// key to allow SSH connections. Note that it explicitly doesn't actually
+// support any Auth mechanisms, so nobody can complete connections, but it will
+// do Key exchange to set up the encrypted conversation.
+// We return the address where the SSH service is listening, and a channel
+// callers must close when they want the service to stop.
+func CreateSSHServer(c *gc.C, privateKeys ...string) (string, chan struct{}) {
```

Comment:
This is quite similar in structure to testTCPServer. Could this take a flag or something to make it not do any key exchange so that it can be used for both SSH and plain TCP testing? (avoiding some duplicated code)